File: aws-apigateway-py-routes/lambdas.py
# Copyright 2016-2021, Pulumi Corporation.
import pulumi
import pulumi_aws as aws

lambda_role = aws.iam.Role('auth-lambda-role',
                           assume_role_policy="""{
        "Version": "2012-10-17",
        "Statement": [
            {
                "Action": "sts:AssumeRole",
                "Principal": {
                    "Service": "lambda.amazonaws.com"
                },
                "Effect": "Allow",
                "Sid": ""
            }
        ]
    }"""
                           )

lambda_role_policy = aws.iam.RolePolicy('auth-lambda-role-policy',
                                        role=lambda_role.id,
                                        policy="""{
        "Version": "2012-10-17",
        "Statement": [{
            "Effect": "Allow",
            "Action": [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents"
            ],
            "Resource": "arn:aws:logs:*:*:*"
        }]
    }"""
                                        )

# Create a Lambda function to validate request authorization
auth_lambda = aws.lambda_.Function("auth-lambda",
                                   role=lambda_role.arn,
                                   runtime=aws.lambda_.Runtime.PYTHON3D8,
                                   code=pulumi.AssetArchive({
                                       ".": pulumi.FileArchive("./authorizer"),
                                   }),
                                   handler="handler.handler",
                                   )

# Create a Lambda function to respond to HTTP requests
hello_handler = aws.lambda_.Function("hello-handler",
                                     role=lambda_role.arn,
                                     runtime=aws.lambda_.Runtime.PYTHON3D8,
                                     code=pulumi.AssetArchive({
                                         ".": pulumi.FileArchive("./handler"),
                                     }),
                                     handler="handler.handler",
                                     )

File: aws/go/aws-apigateway-go-routes/lambdas.go
// Copyright 2016-2021, Pulumi Corporation.
package main

import (
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/iam"
	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/lambda"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func createLambdas(ctx *pulumi.Context) (*lambda.Function, *lambda.Function, error) {
	role, err := iam.NewRole(ctx, "auth-exec-role", &iam.RoleArgs{
		AssumeRolePolicy: pulumi.String(`{
				"Version": "2012-10-17",
				"Statement": [{
					"Sid": "",
					"Effect": "Allow",
					"Principal": {
						"Service": "lambda.amazonaws.com"
					},
					"Action": "sts:AssumeRole"
				}]
			}`),
	})
	if err != nil {
		return nil, nil, err
	}

	logPolicy, err := iam.NewRolePolicy(ctx, "lambda-log-policy", &iam.RolePolicyArgs{
		Role: role.Name,
		Policy: pulumi.String(`{
                "Version": "2012-10-17",
                "Statement": [{
                    "Effect": "Allow",
                    "Action": [
                        "logs:CreateLogGroup",
                        "logs:CreateLogStream",
                        "logs:PutLogEvents"
                    ],
                    "Resource": "arn:aws:logs:*:*:*"
                }]
            }`),
	})
	if err != nil {
		return nil, nil, err
	}

	authLambda, err := lambda.NewFunction(ctx, "auth-lambda", &lambda.FunctionArgs{
		Runtime: lambda.RuntimeGo1dx,
		Code: pulumi.NewAssetArchive(map[string]interface{}{
			".": pulumi.NewFileArchive("./bin/authorizer.zip"),
		}),
		Handler: pulumi.String("authorizer"),
		Role:    role.Arn,
	}, pulumi.DependsOn([]pulumi.Resource{logPolicy}))
	if err != nil {
		return nil, nil, err
	}

	helloHandler, err := lambda.NewFunction(ctx, "hello-handler", &lambda.FunctionArgs{
		Runtime: lambda.RuntimeGo1dx,
		Code: pulumi.NewAssetArchive(map[string]interface{}{
			".": pulumi.NewFileArchive("./bin/handler.zip"),
		}),
		Handler: pulumi.String("handler"),
		Role:    role.Arn,
	}, pulumi.DependsOn([]pulumi.Resource{logPolicy}))
	if err != nil {
		return nil, nil, err
	}
	return authLambda, helloHandler, nil
}
